### PR TITLE
Block task execution on catalog loading flag

### PR DESCRIPTION
**Pull request by jklamer**
## Description
Block task updating on all catalogs being loaded. Currently the second request to task update will always update the task before waiting for the catalogs to be loaded. This will block all threads trying to update the task on waiting for catalogs to load. 

fixes https://github.com/trinodb/trino/issues/20326

## Release notes

( X) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/execution/SqlTaskManager.java [hunk 1]
```diff
@@ -536,11 +536,12 @@ private TaskInfo doUpdateTask(
                     Set<CatalogHandle> catalogHandles = activeCatalogs.stream()
                             .map(CatalogProperties::catalogHandle)
                             .collect(toImmutableSet());
-                    if (sqlTask.setCatalogs(catalogHandles)) {
+                    if (sqlTask.setCatalogs(catalogHandles) || !sqlTask.catalogsLoaded()) {
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/execution/SqlTaskManager.java:539]: dio do we have two use double-check pattern here?
we get `!sqlTask.catalogsLoaded()` and then we take a lock and re-check

the whole complexity could perhaps be hidden inside `ensureCatalogsLoaded()` method.
if it cannot, the code should be commented, documenting why the extra complexity is needed

**Comment by wendigo** [comment]
@jklamer seems that it doesn't work correctly

File: core/trino-main/src/main/java/io/trino/execution/SqlTaskManager.java (hunk 1 above)
Comment (dain) [reply core/trino-main/src/main/java/io/trino/execution/SqlTaskManager.java:539]: @jklamer let's talk about simplifying this code in some follow up work.  I'm thinking we can hoist the read lock around the larger statement, and then move the entire update inside of Task by giving Task access to `connectorServicesProvider`... maybe.  It is just a thought.

File: .github/workflows/ci.yml
Comment (dain) [inline .github/workflows/ci.yml:1]: are we dropping this Commit?

File: core/trino-main/src/main/java/io/trino/connector/WorkerDynamicCatalogManager.java [hunk 2]
```diff
@@ -61,7 +61,7 @@ public class WorkerDynamicCatalogManager
     private final Lock catalogLoadingLock = catalogsLock.readLock();
     private final Lock catalogRemovingLock = catalogsLock.writeLock();
     private final ConcurrentMap<CatalogHandle, CatalogConnector> catalogs = new ConcurrentHashMap<>();
-    private final ExecutorService executor = Executors.newVirtualThreadPerTaskExecutor();
+    private final ExecutorService executor = Executors.newCachedThreadPool();
```
Comment (dain) [inline core/trino-main/src/main/java/io/trino/connector/WorkerDynamicCatalogManager.java:64]: This is unbounded, so we are likely going need some config to limit the size of this.

**Review by dain — COMMENTED** [review]
I have one comment, but looks like there are relevant test failures.

File: .github/workflows/ci.yml
Comment (wendigo) [reply .github/workflows/ci.yml:1]: Yep. I've added this to stress test raptor legacy.

File: core/trino-main/src/main/java/io/trino/execution/SqlTaskManager.java (hunk 1 above)
Comment (jklamer) [reply core/trino-main/src/main/java/io/trino/execution/SqlTaskManager.java:539]: yeah, I think can rework this mechanism. The key complexity is that the catalogs being set in the SqlTask has to be ordered with the prune catalog task in this class. And then internal to the Catalog manager the ensureCatalogLoaded has to be ordered with the pruning as well, but for different reasons.
If the catalog manager was able to add the catalogs to the task we might be able to flatten it all internal to the manager

Comment (findepi) [reply core/trino-main/src/main/java/io/trino/execution/SqlTaskManager.java:539]: what does it mean "has to be ordered with the prune catalog task in this class"?

Comment (dain) [inline core/trino-main/src/main/java/io/trino/execution/SqlTaskManager.java:539]: With this change, I don't think setCatalogs needs to have a return value.  The main trigger is now `catalogsLoaded`, so this code would be something like this:
```suggestion
sqlTask.setCatalogs(catalogHandles);
if ( !sqlTask.catalogsLoaded()) {
```

Comment (dain) [reply core/trino-main/src/main/java/io/trino/execution/SqlTaskManager.java:539]: I don't think "ordered" is the right term here.  To prune catalogs, we must temporarily pause adding catalogs while we compute the which catalogs to prune.

**Review by dain — COMMENTED** [review]
One comment, but otherwise looks good.  When you patch that, can you remove the empty commits?

**Comment by wendigo** [comment]
@dain @jklamer can you work towards thinking this?

**Comment by findepi** [comment]
@jklamer is this PR "fixes https://github.com/trinodb/trino/issues/20326"?

**Comment by jklamer** [comment]
@findepi Correct this PR is designed to fix and disallow the state where https://github.com/trinodb/trino/issues/20326 happens. There should be no creation of task executions until there is positive confirmation that the catalogs have been loaded.

**Comment by findepi** [comment]
@jklamer thanks for confirmation. can you please add "fixes https://github.com/trinodb/trino/issues/20326" in the PR description?

**Comment by wendigo** [comment]
@findepi i've been stress testing that change and it's not bullet proof hence I do not believe it's fixing issue you propose

**Comment by jklamer** [comment]
Yeah I believe this takes the current mechanism as designed to its limits (or I can't think of any more potential issues in the current mechanism). If this still has issues I would like to get our heads together and think about changing the mechanism in a way we're sure will work for us.

**Comment by jklamer** [comment]
3 Test Failures:
Raptor Legacy, lots of noise around shard cleanup, but the test failure that seems to be the issue is:
```
Error:  Tests run: 667, Failures: 0, Errors: 1, Skipped: 129, Time elapsed: 286.8 s <<< FAILURE! -- in io.trino.plugin.raptor.legacy.TestRaptorBucketedConnectorTest
Error:  io.trino.plugin.raptor.legacy.TestRaptorBucketedConnectorTest.testColumnName -- Time elapsed: 32.98 s <<< ERROR!
io.trino.testing.QueryFailedException: Failed to perform metadata operation
	at io.trino.testing.AbstractTestingTrinoClient.execute(AbstractTestingTrinoClient.java:133)
	at io.trino.testing.DistributedQueryRunner.executeInternal(DistributedQueryRunner.java:501)
	at io.trino.testing.DistributedQueryRunner.executeWithPlan(DistributedQueryRunner.java:490)
	at io.trino.testing.QueryAssertions.assertDistributedUpdate(QueryAssertions.java:108)
	at io.trino.testing.QueryAssertions.assertUpdate(QueryAssertions.java:62)
	at io.trino.testing.AbstractTestQueryFramework.assertUpdate(AbstractTestQueryFramework.java:410)
	at io.trino.testing.AbstractTestQueryFramework.assertUpdate(AbstractTestQueryFramework.java:405)
	at io.trino.testing.BaseConnectorTest.testColumnName(BaseConnectorTest.java:5260)
	at io.trino.testing.BaseConnectorTest.testColumnName(BaseConnectorTest.java:5249)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at java.base/java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:194)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:507)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.tryRemoveAndExec(ForkJoinPool.java:1534)
	at java.base/java.util.concurrent.ForkJoinPool.helpJoin(ForkJoinPool.java:2269)
	at java.base/java.util.concurrent.ForkJoinTask.awaitDone(ForkJoinTask.java:495)
	at java.base/java.util.concurrent.ForkJoinTask.join(ForkJoinTask.java:662)
	at java.base/java.util.concurrent.RecursiveAction.exec(RecursiveAction.java:194)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:507)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1491)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:2073)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:2035)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:187)
	Suppressed: java.lang.Exception: SQL: CREATE TABLE tcn_aslasheahu1gqf39(key varchar(50), "a/slash`" varchar(50))
		at io.trino.testing.DistributedQueryRunner.executeInternal(DistributedQueryRunner.java:508)
		... 20 more
Caused by: io.trino.spi.TrinoException: Failed to perform metadata operation
	at io.trino.plugin.raptor.legacy.util.DatabaseUtil.metadataError(DatabaseUtil.java:86)
	at io.trino.plugin.raptor.legacy.util.DatabaseUtil.metadataError(DatabaseUtil.java:91)
	at io.trino.plugin.raptor.legacy.metadata.DatabaseShardManager.createTable(DatabaseShardManager.java:213)
	at io.trino.plugin.raptor.legacy.RaptorMetadata.finishCreateTable(RaptorMetadata.java:711)
	at io.trino.plugin.raptor.legacy.RaptorMetadata.createTable(RaptorMetadata.java:454)
	at io.trino.spi.connector.ConnectorMetadata.createTable(ConnectorMetadata.java:532)
	at io.trino.tracing.TracingConnectorMetadata.createTable(TracingConnectorMetadata.java:394)
	at io.trino.metadata.MetadataManager.createTable(MetadataManager.java:819)
	at io.trino.tracing.TracingMetadata.createTable(TracingMetadata.java:409)
	at io.trino.execution.CreateTableTask.internalExecute(CreateTableTask.java:297)
	at io.trino.execution.CreateTableTask.execute(CreateTableTask.java:126)
	at io.trino.execution.CreateTableTask.execute(CreateTableTask.java:92)
	at io.trino.execution.DataDefinitionExecution.start(DataDefinitionExecution.java:146)
	at io.trino.execution.SqlQueryManager.createQuery(SqlQueryManager.java:272)
	at io.trino.dispatcher.LocalDispatchQuery.startExecution(LocalDispatchQuery.java:145)
	at io.trino.dispatcher.LocalDispatchQuery.lambda$waitForMinimumWorkers$2(LocalDispatchQuery.java:129)
	at io.airlift.concurrent.MoreFutures.lambda$addSuccessCallback$12(MoreFutures.java:570)
	at io.airlift.concurrent.MoreFutures$3.onSuccess(MoreFutures.java:545)
	at com.google.common.util.concurrent.Futures$CallbackListener.run(Futures.java:1137)
	at io.trino.$gen.Trino_testversion____20240409_202519_304.run(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1570)
Caused by: org.jdbi.v3.core.statement.UnableToExecuteStatementException: org.h2.jdbc.JdbcSQLTimeoutException: Timeout trying to lock table "SYS"; SQL statement:
CREATE TABLE x_shards_t93 (
  shard_id BIGINT NOT NULL,
  shard_uuid BINARY(16) NOT NULL,
  node_ids VARBINARY(128) NOT NULL,
  c1_min varbinary(100),
  c1_max varbinary(100),
  c2_min varbinary(100),
  c2_max varbinary(100),
  PRIMARY KEY (node_ids, shard_uuid),
  UNIQUE (shard_id),
  UNIQUE (shard_uuid),
  UNIQUE (node_ids, shard_id, shard_uuid)
) [50200-224] [statement:"CREATE TABLE x_shards_t93 (
  shard_id BIGINT NOT NULL,
  shard_uuid BINARY(16) NOT NULL,
  node_ids VARBINARY(128) NOT NULL,
  c1_min varbinary(100),
  c1_max varbinary(100),
  c2_min varbinary(100),
  c2_max varbinary(100),
  PRIMARY KEY (node_ids, shard_uuid),
  UNIQUE (shard_id),
  UNIQUE (shard_uuid),
  UNIQUE (node_ids, shard_id, shard_uuid)
)", arguments:{positional:{}, named:{}, finder:[]}]
	at org.jdbi.v3.core.statement.SqlStatement.internalExecute(SqlStatement.java:1815)
	at org.jdbi.v3.core.result.ResultProducers.lambda$returningUpdateCount$0(ResultProducers.java:49)
	at org.jdbi.v3.core.statement.Update.execute(Update.java:69)
	at org.jdbi.v3.core.statement.Update.execute(Update.java:57)
	at org.jdbi.v3.core.Handle.execute(Handle.java:354)
	at org.jdbi.v3.core.Handle.execute(Handle.java:367)
	at io.trino.plugin.raptor.legacy.metadata.DatabaseShardManager.createTable(DatabaseShardManager.java:210)
	... 20 more
Caused by: org.h2.jdbc.JdbcSQLTimeoutException: Timeout trying to lock table "SYS"; SQL statement:
CREATE TABLE x_shards_t93 (
  shard_id BIGINT NOT NULL,
  shard_uuid BINARY(16) NOT NULL,
  node_ids VARBINARY(128) NOT NULL,
  c1_min varbinary(100),
  c1_max varbinary(100),
  c2_min varbinary(100),
  c2_max varbinary(100),
  PRIMARY KEY (node_ids, shard_uuid),
  UNIQUE (shard_id),
  UNIQUE (shard_uuid),
  UNIQUE (node_ids, shard_id, shard_uuid)
) [50200-224]
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:572)
	at org.h2.message.DbException.getJdbcSQLException(DbException.java:489)
	at org.h2.message.DbException.get(DbException.java:223)
	at org.h2.message.DbException.get(DbException.java:199)
	at org.h2.mvstore.db.MVTable.doLock1(MVTable.java:230)
	at org.h2.mvstore.db.MVTable.lock(MVTable.java:189)
	at org.h2.engine.Database.lockMeta(Database.java:778)
	at org.h2.command.ddl.CreateTable.update(CreateTable.java:85)
	at org.h2.command.CommandContainer.update(CommandContainer.java:169)
	at org.h2.command.Command.executeUpdate(Command.java:256)
	at org.h2.jdbc.JdbcPreparedStatement.execute(JdbcPreparedStatement.java:265)
	at org.jdbi.v3.core.statement.SqlLoggerUtil.wrap(SqlLoggerUtil.java:32)
	at org.jdbi.v3.core.statement.SqlStatement.internalExecute(SqlStatement.java:1813)
	... 26 more
```
Test-Pinot: timed out
Test-jdbc-compatibility:
timed out waiting for test to run after 8 minutes. Looks like the 8 minutes involved some docker image pulling and trino starting.

**Comment by dain** [comment]
These failures seem unrelated to me.  I just restarted the failed tests to see if they go away.

**Pull request merged**